### PR TITLE
Add "Check for updates" button

### LINKED ISSUE #8947: Add manual "Check for updates" button in settings
**Describe the new feature:** 
-------------------------------------
Some times the internet connection are unstable or too slow that take too long time to update, least I am.
The game are not trying to update everytime when start if you close it before downloading update finish

Please add a _**Check Update**_ button in settings below _**Open Osu! Folder**_ for manually start update or check update everytime the game starts

**Proposal designs of the feature:**
-------------------------------------

Add a _**Check update**_ button in settings below **_Open Osu! Folder_**
Check for update everytime the game starts

### PR DESCRIPTION
**Pull request by Craftplacer**
Closes #8947
- [x] Depends on #9002

### CODE REVIEW

File: osu.Game/Updater/UpdateManager.cs
Comment (bdach) [inline osu.Game/Updater/UpdateManager.cs:50]: Why is this being logged at all?

File: osu.Game/OsuGame.cs
Comment (bdach) [inline osu.Game/OsuGame.cs:613]: Is there any particular reason for this location of this load?

File: osu.Game/OsuGame.cs
Comment (bdach) [inline osu.Game/OsuGame.cs:780]: I find this a little dangerous. I assume you've changed the `Cache` to `CacheAs` so that components aren't cached with their most derived type (returned at runtime), but with their *declaration* type as written in the field/local. This is an extremely subtle change that can cause regressions if any of the types anywhere don't match - did you make sure that *none* of the single-file-loaded cached dependencies break?

File: osu.Game/Overlays/Settings/Sections/General/UpdateSettings.cs
Comment (bdach) [inline osu.Game/Overlays/Settings/Sections/General/UpdateSettings.cs:39]: You could just as well do

```suggestion
            Add(new SettingsButton
            {
                Text = "Check for updates",
                Action = updateManager?.CheckForUpdate,
                Enabled = { Value = game?.IsDeployedBuild ?? false }
            });
```

to eliminate the `if` null-checks.

File: osu.Game/OsuGame.cs
Comment (Craftplacer) [reply osu.Game/OsuGame.cs:613]: UpdateManager depends on NotificationOverlay. SettingsOverlay has UpdateSettings which depends on UpdateManager, so it has to be "sandwiched".

File: osu.Game/Updater/UpdateManager.cs
Comment (Craftplacer) [reply osu.Game/Updater/UpdateManager.cs:50]: I felt this is the right thing to do. You don't want to silently do nothing, because you missed something or expected something else.

Comment (bdach) [reply osu.Game/Updater/UpdateManager.cs:50]: Doesn't seem very useful. There is no practical difference to the user between doing nothing and logging this (default log level is verbose, so it won't show up anywhere unless you manually check the log). I'd argue that in the cases that would log this, the button should be either disabled or not even visible to users.

I'm pretty sure the only platform that uses the base `UpdateManager` directly is iOS, on which the update button is probably pretty useless anyway in the long run due to the app store existing (I presume testflight also handles app updates, but I wouldn't know).

File: osu.Game/OsuGame.cs
Comment (bdach) [reply osu.Game/OsuGame.cs:613]: The `UpdateManager -> SettingsOverlay` dependency should be annotated with an inline comment for posterity in that case, albeit I don't really like how interdependent everything is starting to become.

File: osu.Game/OsuGame.cs
Comment (Craftplacer) [reply osu.Game/OsuGame.cs:780]: Yes, everything is working.

File: osu.Game/Updater/UpdateManager.cs
Comment (Craftplacer) [reply osu.Game/Updater/UpdateManager.cs:50]: I'm going to keep the log message for purely debugging purposes. But I'm going to hide the button if we're having the base class.

File: osu.Game/OsuGame.cs
Comment (bdach) [reply osu.Game/OsuGame.cs:780]: If I may ask, how did you determine that? To be able to say that with certainty would require to inspect every BDL/`[Resolved]` member that uses the types cached here and make sure none of them use a more-derived type than the one declared in this class.

Comment (iiSaLMaN) [reply osu.Game/OsuGame.cs:780]: That's the same as checking for every `loadComponentSingleFile<T>` call to look for ones that have the instantiated type more-derived than the `T` generic type, in which case I couldn't find one ([other than the already-resolved case in the branch](https://github.com/ppy/osu/pull/8965/commits/c025814f403dc5fcf4f07342791387986c526809#diff-4af0c24c8977f999b5986c46cc749ec3R93)), that would mean a confirmed yes.

Moreover, caching components as `T` type IMO is better than most-derived type for this case of loading game components, just like what's being done here (with `CreateUpdateManager()` instance cached as `UpdateManager`), there may be specific components with platform-specific subclasses that need to be cached as their one shared base class to allow resolving them throughout the game.

Comment (bdach) [reply osu.Game/OsuGame.cs:780]: I don't necessarily disagree with the content of the change, I mostly wanted to make sure it was tested as safe.

File: osu.Game/Overlays/Settings/Sections/General/UpdateSettings.cs
Comment (bdach) [inline osu.Game/Overlays/Settings/Sections/General/UpdateSettings.cs:29]: Not sure I like this relying on types this much. I'd probably add a `public virtual bool CanPerformUpdate { get; }` on the update manager (`=> false` in the base, `=> true` in the derived).

Comment (Craftplacer) [reply osu.Game/Overlays/Settings/Sections/General/UpdateSettings.cs:29]: I believe doing a virtual property is more flexible, so I'm going with that.
Also the default implementation of `CreateUpdateManager()` in `osu.Game.OsuGame`
 passes no arguments to the constructor of `UpdateManager`, and it'd be better to keep it at that.

File: osu.Game/Overlays/Settings/Sections/General/UpdateSettings.cs
Comment (peppy) [inline osu.Game/Overlays/Settings/Sections/General/UpdateSettings.cs:29]:
```suggestion
            if (updateManager?.CanPerformUpdate == true)
```

File: osu.Game/Updater/UpdateManager.cs
Comment (peppy) [inline osu.Game/Updater/UpdateManager.cs:20]: no xmldoc. what does this even mean?

File: osu.Game/Updater/SimpleUpdateManager.cs
Comment (peppy) [inline osu.Game/Updater/SimpleUpdateManager.cs:38]: public method that bypasses the `IsDeployedBuild` check? should probably be protected at most.

File: osu.Game/OsuGame.cs
Comment (peppy) [inline osu.Game/OsuGame.cs:780]: this should be its own PR with reasoning provided

File: osu.Game/OsuGame.cs
Comment (peppy) [inline osu.Game/OsuGame.cs:613]: are you sure this doesn't regress anything?

File: osu.Game/OsuGame.cs
Comment (peppy) [inline osu.Game/OsuGame.cs:93]: How is this even related to this PR? Should not be here.

File: osu.Game/Updater/UpdateManager.cs
Comment (peppy) [reply osu.Game/Updater/UpdateManager.cs:20]: why is this not including the IsDeployedBuild check?

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/OsuGame.cs
Comment (Craftplacer) [reply osu.Game/OsuGame.cs:613]: image
osu!'s update check at startup still works, so this is a yes from me.

File: osu.Game/Updater/UpdateManager.cs
Comment (Craftplacer) [reply osu.Game/Updater/UpdateManager.cs:20]: After I committed it, I myself doubted the name of this property, a more appropriate name would be "CanCheckForUpdate".  That name would make more sense and is self-explanatory.

Comment (peppy) [reply osu.Game/Updater/UpdateManager.cs]: but why is IsDeployedBuild not included and how does that check differ from the one you added?

File: osu.Game/OsuGame.cs
Comment (bdach) [reply osu.Game/OsuGame.cs:93]: It's a corollary of the `Cache` -> `CacheAs` change [as I outlined here](https://github.com/ppy/osu/pull/8965#discussion_r421827375). I too dislike the subtlety of this.

File: osu.Game/OsuGame.cs
Comment (peppy) [reply osu.Game/OsuGame.cs:780]: @Craftplacer are you going to split this out?

**Comment by Craftplacer** [comment]
The latest commit intentionally breaks the current code (as I transferred the changes over there, which makes merging easier), while #9002 is pending to be discussed and accepted.

File: osu.Game/Updater/UpdateManager.cs
Comment (peppy) [inline osu.Game/Updater/UpdateManager.cs:55]: This is wrong

File: osu.Game/Updater/SimpleUpdateManager.cs
Comment (peppy) [inline osu.Game/Updater/SimpleUpdateManager.cs:22]: why is this not IsDeployedBuild still?

File: osu.Game/Overlays/Settings/Sections/General/UpdateSettings.cs
Comment (peppy) [inline osu.Game/Overlays/Settings/Sections/General/UpdateSettings.cs:35]: what?

File: osu.Desktop/Updater/SquirrelUpdateManager.cs
Comment (peppy) [inline osu.Desktop/Updater/SquirrelUpdateManager.cs:25]: why is this not IsDeployedBuild still?

**Review by peppy — DISMISSED** [review]
I think you missed the point.

**Comment by peppy** [comment]
@Craftplacer are you planning on revisiting this?

**Comment by Craftplacer** [comment]
> @Craftplacer are you planning on revisiting this?

This weekend most likely. I was waiting for contributions of another contributor that didn't came out to be.

File: osu.Game/Overlays/Settings/Sections/General/UpdateSettings.cs
Comment (frenzibyte) [inline osu.Game/Overlays/Settings/Sections/General/UpdateSettings.cs]: What is this for?

File: osu.Desktop/Updater/SquirrelUpdateManager.cs
Comment (frenzibyte) [inline osu.Desktop/Updater/SquirrelUpdateManager.cs:38]: The base already calls this, you obviously don't want to check twice?

File: osu.Game/Updater/UpdateManager.cs
Comment (frenzibyte) [inline osu.Game/Updater/UpdateManager.cs:23]: This will fail on iOS, maybe do:
```csharp
public bool CanPerformUpdate => game.IsDeployedBuild && CanCheckForUpdate;

protected virtual bool CanCheckForUpdate => false;
```
but not sure about the naming though, may want a more saner naming here

File: osu.Game/Updater/UpdateManager.cs
Comment (frenzibyte) [inline osu.Game/Updater/UpdateManager.cs]: what is this doing here, should stay in `LoadComplete`?

File: osu.Game/Updater/UpdateManager.cs
Comment (frenzibyte) [inline osu.Game/Updater/UpdateManager.cs:50]: No, don't silently return when you shouldn't be called, throw `NotSupportedException` here.

**Review by frenzibyte — CHANGES_REQUESTED** [review]

File: osu.Game/Updater/UpdateManager.cs
Comment (peppy) [reply osu.Game/Updater/UpdateManager.cs:50]: reasoning?

Comment (frenzibyte) [reply osu.Game/Updater/UpdateManager.cs:50]: `CheckForUpdateAsync` shouldn't be called if the `UpdateManager` itself isn't capable of doing it, that should be identified as a bug.

Comment (peppy) [reply osu.Game/Updater/UpdateManager.cs:50]: I don't follow. It's a public method. Are you expecting every usage to first check the bool? That sounds like really bad API. Why don't we just make UpdateManager null and throw a nullref then? Isn't the point of making it abstract to simplify external implementation?

File: osu.Game/Updater/UpdateManager.cs
Comment (Craftplacer) [reply osu.Game/Updater/UpdateManager.cs]: You highlighted two methods, what do you mean by "this"?

File: osu.Game/Updater/UpdateManager.cs
Comment (Craftplacer) [reply osu.Game/Updater/UpdateManager.cs:23]: Why would this be failing, because IsDeployedBuild is false on iOS?

File: osu.Game/Overlays/Settings/Sections/General/UpdateSettings.cs
Comment (Craftplacer) [reply osu.Game/Overlays/Settings/Sections/General/UpdateSettings.cs]: I'm pretty sure that this is another PR that got merged (it doesn't get highlighted as my change on Files Changed).

File: osu.Game/Updater/UpdateManager.cs
Comment (bdach) [reply osu.Game/Updater/UpdateManager.cs:23]: Yeah, I think as long as iOS reads the assembly version properly this should work with the simple update manager?

Comment (frenzibyte) [reply osu.Game/Updater/UpdateManager.cs:23]: iOS is on a completely different version publishing system (by TestFlight), hence the reason if you open `OsuGameIOS` you would find `SimpleUpdateManager` isn't used there.

With this staying as-is, `Check for update` button will be enabled for iOS, which we're likely not expecting here given how it's published.

File: osu.Game/Updater/UpdateManager.cs
Comment (frenzibyte) [reply osu.Game/Updater/UpdateManager.cs:50]: Right, you have a point there. It's fine in this way then.

File: osu.Game/Overlays/Settings/Sections/General/UpdateSettings.cs
Comment (frenzibyte) [reply osu.Game/Overlays/Settings/Sections/General/UpdateSettings.cs]: Weird, how did this got displayed to me as if it was your change o_O

File: osu.Game/Updater/UpdateManager.cs
Comment (frenzibyte) [reply osu.Game/Updater/UpdateManager.cs]: Another weird thing by GItHub, anyways why is the version finalizing logic inside `InternalCheckForUpdateAsync`, it should stay where it is at `LoadComplete`?

File: osu.Game/Updater/UpdateManager.cs
Comment (bdach) [reply osu.Game/Updater/UpdateManager.cs:23]: Fair. I got my virtual implementations confused.

File: osu.Game/Updater/UpdateManager.cs
Comment (Craftplacer) [reply osu.Game/Updater/UpdateManager.cs:50]: > I don't follow. It's a public method. Are you expecting every usage to first check the bool? That sounds like really bad API. Why don't we just make UpdateManager null and throw a nullref then? Isn't the point of making it abstract to simplify external implementation?

I thought about doing that too, but UpdateManager seems to go beyond of just checking updates.

At this point I would suggest splitting UpdateManager into 2 classes for their own functionality, that being a class for checking for new updates, and another one for doing the startup work when a new update has been applied.

This way we can stick to abstracting the checking process and make it null on not deployed builds/mobile versions.

File: osu.Game/Updater/UpdateManager.cs
Comment (Craftplacer) [reply osu.Game/Updater/UpdateManager.cs]: I mixed up the "Check what's new" code as another checking code, I'll be reverting this the next push.

File: osu.Game/Updater/UpdateManager.cs
Comment (bdach) [reply osu.Game/Updater/UpdateManager.cs:50]: I can't see the gain of splitting the two things. Sounds like extra complexity for not much gain.

**Comment by peppy** [comment]
I'm taking this PR over. There is a lack of effort or testing and reviewing further is not efficient.

**Comment by peppy** [comment]
- Fixed naming
- Fixed button not disabling while check is in progress
- Fixed update being run twice every startup
- Fixed no check for concurrent update checks

File: osu.Game/Updater/UpdateManager.cs
Comment (peppy) [reply osu.Game/Updater/UpdateManager.cs]: Fix in a slightly different way.

File: osu.Game/Updater/UpdateManager.cs
Comment (peppy) [inline osu.Game/Updater/UpdateManager.cs:42]: what

**Comment by peppy** [comment]
@smoogipoo requesting your thorough check of this (although there's not much of it left).

File: osu.Game/Overlays/Settings/Sections/General/UpdateSettings.cs
Comment (smoogipoo) [inline osu.Game/Overlays/Settings/Sections/General/UpdateSettings.cs:40]: Deserves a null check (`CanBeNull = true` on the attribute).

File: osu.Game/Updater/UpdateManager.cs
Comment (smoogipoo) [inline osu.Game/Updater/UpdateManager.cs:70]: Safest would be to store the above to a local variable and await on the local variable here. T1 could be executing the statement above, and T2 could be blocked on the statement below, after which T2 will unblock and set null and T1 will try `await null;`

**Review by smoogipoo — CHANGES_REQUESTED** [review]

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**
